src/proof.rs

```diff
@@ -19,23 +26,45 @@ use crate::test_common::{get_vec_tree_from_slice, Item, XOR128};
 ///
 /// Proof validation is positioned hash against lemma path to match root hash.
 #[derive(Debug, Clone, Eq, PartialEq)]
+/// U is the default tree arity (U2 = binary)
 pub struct Proof<T: Eq + Clone + AsRef<[u8]>, U: Unsigned = U2> {
+    // Optional proofs at immediate lower level from current.  Should
+    // be None at the base layer.
+    sub_tree_proof: Option<Box<Proof<T, U>>>,
+
+    top_layer_nodes: usize,      // arity of top layer
+    sub_tree_layer_nodes: usize, // arity of sub-tree layer
+
     lemma: Vec<T>,
-    path: Vec<usize>,   // branch index
+    path: Vec<usize>, // branch index
+
     _u: PhantomData<U>, // number of branches per node
 }
 
 impl<T: Eq + Clone + AsRef<[u8]>, U: Unsigned> Proof<T, U> {
     /// Creates new MT inclusion proof
-    pub fn new(lemma: Vec<T>, path: Vec<usize>) -> Result<Proof<T, U>> {
-        ensure!(lemma.len() > 2, "Invalid lemma length (short)");
-        ensure!(
-            lemma.len() == get_merkle_proof_lemma_len(path.len() + 1, U::to_usize()),
-            "Invalid lemma length"
-        );
+    pub fn new(
+        sub_tree_proof: Option<Box<Proof<T, U>>>,
+        lemma: Vec<T>,
+        path: Vec<usize>,
+        top_layer_nodes: usize,
+        sub_tree_layer_nodes: usize,
```

Comment:
I would pass these values as type arguments, rather than regular values, even if they are not stored as such, as this makes the API a little safer.

```rust
fn new<TopLayerNodes: Unsigned, SubTreeLayerNodes: Unsigned>(sub_tree_proof, lemma, path) -> Result<Proof<T, U>> {
}
```